File: drivers/cpp/src/redwire/codec.cpp
#include "reddb/redwire/codec.hpp"
#include "reddb/errors.hpp"

#include <cstring>

#if REDDB_HAS_ZSTD
#include <zstd.h>
#endif

namespace reddb::redwire {

bool zstd_available() noexcept {
#if REDDB_HAS_ZSTD
    return true;
#else
    return false;
#endif
}

namespace {

#if REDDB_HAS_ZSTD
std::vector<uint8_t> zstd_compress(const std::vector<uint8_t>& src, int level) {
    size_t bound = ZSTD_compressBound(src.size());
    std::vector<uint8_t> out(bound);
    size_t n = ZSTD_compress(out.data(), bound, src.data(), src.size(), level);
    if (ZSTD_isError(n)) {
        throw RedDBError(ErrorCode::Protocol,
                         std::string("zstd compress: ") + ZSTD_getErrorName(n));
    }
    out.resize(n);
    return out;
}

std::vector<uint8_t> zstd_decompress(const uint8_t* src, size_t n) {
    // Try the framed size first; fall back to streaming for unknown sizes.
    unsigned long long size = ZSTD_getFrameContentSize(src, n);
    if (size != ZSTD_CONTENTSIZE_ERROR && size != ZSTD_CONTENTSIZE_UNKNOWN) {
        std::vector<uint8_t> out(static_cast<size_t>(size));
        size_t got = ZSTD_decompress(out.data(), out.size(), src, n);
        if (ZSTD_isError(got)) {
            throw RedDBError(ErrorCode::Protocol,
                             std::string("zstd decompress: ") + ZSTD_getErrorName(got));
        }
        out.resize(got);
        return out;
    }
    // Streaming decode.
    std::vector<uint8_t> out;
    ZSTD_DStream* ds = ZSTD_createDStream();
    if (!ds) {
        throw RedDBError(ErrorCode::Protocol, "zstd: createDStream failed");
    }
    ZSTD_initDStream(ds);
    ZSTD_inBuffer in{src, n, 0};
    uint8_t buf[8192];
    while (in.pos < in.size) {
        ZSTD_outBuffer ob{buf, sizeof(buf), 0};
        size_t r = ZSTD_decompressStream(ds, &ob, &in);
        if (ZSTD_isError(r)) {
            ZSTD_freeDStream(ds);
            throw RedDBError(ErrorCode::Protocol,
                             std::string("zstd decompress: ") + ZSTD_getErrorName(r));
        }
        out.insert(out.end(), buf, buf + ob.pos);
        if (r == 0) break;
    }
    ZSTD_freeDStream(ds);
    return out;
}
#endif

} // namespace

std::vector<uint8_t> encode_frame(const Frame& frame, int zstd_level) {
    std::vector<uint8_t> body = frame.payload;
    Flags flags = frame.flags;

    if (flags.contains(Flags::COMPRESSED)) {
#if REDDB_HAS_ZSTD
        body = zstd_compress(frame.payload, zstd_level);
#else
        (void)zstd_level;
        throw RedDBError(ErrorCode::CompressedButNoZstd,
                         "frame requested COMPRESSED but driver was built without zstd");
#endif
    }

    uint32_t total = static_cast<uint32_t>(FRAME_HEADER_SIZE + body.size());
    if (total > MAX_FRAME_SIZE) {
        throw RedDBError(ErrorCode::Protocol,
                         "encoded frame exceeds MAX_FRAME_SIZE (16 MiB)");
    }

    FrameHeader h;
    h.length = total;
    h.kind = frame.kind;
    h.flags = flags;
    h.stream_id = frame.stream_id;
    h.correlation_id = frame.correlation_id;

    std::vector<uint8_t> out(total);
    encode_header(h, out.data());
    if (!body.empty()) {
        std::memcpy(out.data() + FRAME_HEADER_SIZE, body.data(), body.size());
    }
    return out;
}

std::pair<Frame, size_t> decode_frame(const uint8_t* bytes, size_t len) {
    if (len < FRAME_HEADER_SIZE) {
        throw RedDBError(ErrorCode::Protocol, "frame header truncated");
    }
    FrameHeader h = decode_header(bytes);
    if (h.length < FRAME_HEADER_SIZE || h.length > MAX_FRAME_SIZE) {
        throw RedDBError(ErrorCode::Protocol,
                         "frame length invalid: " + std::to_string(h.length));
    }
    if (len < h.length) {
        throw RedDBError(ErrorCode::Protocol, "frame payload truncated");
    }
    if (!is_known_kind(static_cast<uint8_t>(h.kind))) {
        throw RedDBError(ErrorCode::Protocol,
                         "unknown message kind 0x" +
                         [&]() {
                             char b[3];
                             snprintf(b, sizeof(b), "%02x", static_cast<uint8_t>(h.kind));
                             return std::string(b);
                         }());
    }
    if ((h.flags.bits & ~KNOWN_FLAGS) != 0) {
        throw RedDBError(ErrorCode::Protocol,
                         "unknown flag bits 0x" +
                         [&]() {
                             char b[3];
                             snprintf(b, sizeof(b), "%02x", h.flags.bits);
                             return std::string(b);
                         }());
    }

    size_t payload_len = h.length - FRAME_HEADER_SIZE;
    const uint8_t* on_wire = bytes + FRAME_HEADER_SIZE;

    Frame frame;
    frame.kind = h.kind;
    frame.flags = h.flags;
    frame.stream_id = h.stream_id;
    frame.correlation_id = h.correlation_id;

    if (h.flags.contains(Flags::COMPRESSED)) {
#if REDDB_HAS_ZSTD
        frame.payload = zstd_decompress(on_wire, payload_len);
#else
        throw RedDBError(ErrorCode::CompressedButNoZstd,
                         "frame is COMPRESSED but driver was built without zstd");
#endif
    } else {
        frame.payload.assign(on_wire, on_wire + payload_len);
    }
    return {std::move(frame), static_cast<size_t>(h.length)};
}

} // namespace reddb::redwire

File: drivers/zig/src/redwire/codec.zig
// Frame encode/decode. Compression is opt-in and driven by the
// COMPRESSED flag bit. When the build was configured without zstd
// support, encoding a compressed frame falls back to plaintext
// (matching the JS driver's behaviour) and decoding a compressed
// frame returns `error.CompressedButNoZstd`.

const std = @import("std");
const build_options = @import("build_options");

const frame = @import("frame.zig");
pub const FRAME_HEADER_SIZE = frame.FRAME_HEADER_SIZE;
pub const MAX_FRAME_SIZE = frame.MAX_FRAME_SIZE;
pub const KNOWN_FLAGS = frame.KNOWN_FLAGS;
pub const Flags = frame.Flags;
pub const MessageKind = frame.MessageKind;
pub const Frame = frame.Frame;

const c_zstd = if (build_options.enable_zstd)
    @cImport({
        @cInclude("zstd.h");
    })
else
    struct {};

/// Encode a frame into a freshly-allocated slice. Caller frees.
pub fn encodeFrame(allocator: std.mem.Allocator, f: Frame) ![]u8 {
    var on_wire: []const u8 = f.payload;
    var on_wire_owned: ?[]u8 = null;
    defer if (on_wire_owned) |o| allocator.free(o);

    var out_flags: u8 = f.flags & KNOWN_FLAGS;

    if ((out_flags & Flags.COMPRESSED) != 0) {
        if (build_options.enable_zstd) {
            const compressed = try zstdCompress(allocator, f.payload);
            on_wire_owned = compressed;
            on_wire = compressed;
        } else {
            // No zstd — peel the COMPRESSED flag back off so the
            // peer doesn't try to decompress what we ship plain.
            out_flags &= ~@as(u8, Flags.COMPRESSED);
        }
    }

    const total: usize = FRAME_HEADER_SIZE + on_wire.len;
    if (total > MAX_FRAME_SIZE) return error.FrameTooLarge;

    var buf = try allocator.alloc(u8, total);
    errdefer allocator.free(buf);
    frame.writeHeader(
        buf[0..FRAME_HEADER_SIZE],
        @intCast(total),
        f.kind,
        out_flags,
        f.stream_id,
        f.correlation_id,
    );
    @memcpy(buf[FRAME_HEADER_SIZE..total], on_wire);
    return buf;
}

pub const Decoded = struct {
    frame: Frame,
    consumed: usize,
    /// When non-null, `frame.payload` points into this allocation
    /// and the caller is expected to free it via `deinit`.
    owned_payload: ?[]u8 = null,

    pub fn deinit(self: *Decoded, allocator: std.mem.Allocator) void {
        if (self.owned_payload) |p| {
            allocator.free(p);
            self.owned_payload = null;
        }
    }
};

/// Decode a frame from `bytes`. Returns the parsed frame plus how
/// many bytes were consumed. The returned `Decoded.frame.payload`
/// either borrows from `bytes` (when not compressed) or points into
/// a fresh allocation owned by `Decoded.owned_payload`.
pub fn decodeFrame(allocator: std.mem.Allocator, bytes: []const u8) !Decoded {
    if (bytes.len < FRAME_HEADER_SIZE) return error.FrameTruncated;
    const header = try frame.readHeader(bytes[0..FRAME_HEADER_SIZE]);
    try frame.validateHeader(header);
    if (bytes.len < header.length) return error.FrameTruncated;

    const kind = MessageKind.fromByte(header.kind) orelse return error.FrameUnknownKind;
    const payload_slice = bytes[FRAME_HEADER_SIZE..header.length];

    var out_payload: []const u8 = payload_slice;
    var owned: ?[]u8 = null;
    if ((header.flags & Flags.COMPRESSED) != 0) {
        if (!build_options.enable_zstd) return error.CompressedButNoZstd;
        const plain = try zstdDecompress(allocator, payload_slice);
        owned = plain;
        out_payload = plain;
    }

    return Decoded{
        .frame = .{
            .kind = kind,
            .flags = header.flags,
            .stream_id = header.stream_id,
            .correlation_id = header.correlation_id,
            .payload = out_payload,
        },
        .consumed = header.length,
        .owned_payload = owned,
    };
}

// ---------------------------------------------------------------------------
// zstd glue. Only compiled when build_options.enable_zstd is true.
// ---------------------------------------------------------------------------

fn zstdCompress(allocator: std.mem.Allocator, src: []const u8) ![]u8 {
    if (!build_options.enable_zstd) unreachable;
    const bound = c_zstd.ZSTD_compressBound(src.len);
    const dst = try allocator.alloc(u8, bound);
    errdefer allocator.free(dst);
    const level: c_int = blk: {
        const env = std.posix.getenv("RED_REDWIRE_ZSTD_LEVEL") orelse break :blk 1;
        break :blk std.fmt.parseInt(c_int, env, 10) catch 1;
    };
    const written = c_zstd.ZSTD_compress(
        dst.ptr,
        bound,
        src.ptr,
        src.len,
        level,
    );
    if (c_zstd.ZSTD_isError(written) != 0) return error.DecompressFailed;
    return allocator.realloc(dst, written);
}

fn zstdDecompress(allocator: std.mem.Allocator, src: []const u8) ![]u8 {
    if (!build_options.enable_zstd) unreachable;
    // ZSTD_getFrameContentSize gives the exact uncompressed size or
    // a sentinel when the frame doesn't carry it; in the latter
    // case we grow incrementally via streaming. Engine ships
    // single-frame zstd so the fast path covers it.
    const announced = c_zstd.ZSTD_getFrameContentSize(src.ptr, src.len);
    const ZSTD_CONTENTSIZE_ERROR: u64 = @bitCast(@as(i64, -2));
    const ZSTD_CONTENTSIZE_UNKNOWN: u64 = @bitCast(@as(i64, -1));
    if (announced == ZSTD_CONTENTSIZE_ERROR) return error.DecompressFailed;
    if (announced == ZSTD_CONTENTSIZE_UNKNOWN) {
        // Grow buffer until decompression succeeds — capped at MAX_FRAME_SIZE.
        var cap: usize = @max(src.len * 4, 4096);
        while (cap <= MAX_FRAME_SIZE) : (cap *= 2) {
            const dst = try allocator.alloc(u8, cap);
            const wrote = c_zstd.ZSTD_decompress(dst.ptr, cap, src.ptr, src.len);
            if (c_zstd.ZSTD_isError(wrote) != 0) {
                allocator.free(dst);
                continue;
            }
            return allocator.realloc(dst, wrote);
        }
        return error.DecompressFailed;
    }
    const dst = try allocator.alloc(u8, @intCast(announced));
    errdefer allocator.free(dst);
    const wrote = c_zstd.ZSTD_decompress(dst.ptr, dst.len, src.ptr, src.len);
    if (c_zstd.ZSTD_isError(wrote) != 0) return error.DecompressFailed;
    return allocator.realloc(dst, wrote);
}
